src/pint/models/absolute_phase.py

```diff
@@ -108,16 +108,28 @@ def get_TZR_toa(self, toas):
         #     ephem=toas.ephem,
         #     planets=toas.planets,
         # )
-        tz = toa.get_TOAs_array(
-            (self.TZRMJD.quantity.jd1 - 2400000.5, self.TZRMJD.quantity.jd2),
-            obs=self.TZRSITE.value,
-            freqs=self.TZRFRQ.quantity,
-            include_bipm=clkc_info["include_bipm"],
-            include_gps=clkc_info["include_gps"],
-            ephem=toas.ephem,
-            planets=toas.planets,
-            tzr=True,
-        )
+        if self.TZRSITE.value != "@" and self.TZRSITE.value.lower() != "bat":
+            tz = toa.get_TOAs_array(
```

Comment:
The barycenter has more aliases than "@" and bat. It may be better to change this comparison to
`self.TZRSITE.value.lower() not in [get_observatory("@" ).name] + get_observatory("@" ).aliases`

Reply:
Also, please add a similar condition for the geocenter ("geo") as well.

Reply:
Hi. In tempo2's https://bitbucket.org/psrsoft/tempo2/src/master/refphs.C, this is what it does:
```
 21             if (obs->telID[0]=='@'            /* Have barycentric arrival time */
 22                     || strcasecmp(obs->telID,"bat")==0)
 23             {
 24                 obs->clockCorr=0;  /* therefore don't do clock corrections */
 25                 obs->delayCorr=0;
 26             }
 27             else if (strcmp(obs->telID,"STL")==0)
 28             {
 29                 obs->clockCorr=0;  /* don't do clock corrections */
 30                 obs->delayCorr=1;
 31             }
 32             else if (strcmp(obs->telID,"STL_FBAT")==0)
 33             {
 34                 obs->clockCorr=0;  /* don't do clock corrections */
 35                 obs->delayCorr=1;
 36             }
 37             else
 38             {
 39                 obs->clockCorr=1;
 40                 obs->delayCorr=1;
 41             }
```
It doesn't involve geocentric TOAs at all, and indeed, a par file with "COE" as TZRSITE will get the clock corrections in both PINT and tempo2. If we now disable them in PINT, I am afraid we will end up in the mirrored situation in which PINT doesn't apply the clock corrections to geocentric TZRMJD, whereas tempo2 does, producing again absolute phase differences between the two packages.

Regarding the aliases for "@" and "BAT", this can be easily changed. I have little experience with pytest, but I can try and build a test by looking at other ones in the tests folder.

Reply:
OK... Then you can ignore the part about the geocenter.

About STL and STL_FBAT, I don't know what they are. They are not supported by PINT in any case.

Reply:
I dug into this a bit more. The `include_bipm` and `include_gps` attributes of the `BarycenterObs` class are explicitly set to False in https://github.com/nanograv/PINT/blob/master/src/pint/observatory/special_locations.py

So this bug should not have occurred in the first place. 

I think the bug is in the `get_observatory()` function. When I do `print(get_observatory("@")).include_bipm`, I get True. This is because the `get_observatory()` function wrongly sets this from default arguments. I think the correction should be in `get_observatory()`. This special case treatment is not actually needed.

Reply:
I see! So the check for the barycenter in absolute_phase.py (and toa.py) is unnecessary, provided it is fixed in get_observatory(), am I correct?

Reply:
I think so.

Reply:
Ok, so what about this replacement in get_observatory()?
```
488    site = Observatory.get(name)
489    site.include_gps = include_gps
490    site.include_bipm = include_bipm
491    site.bipm_version = bipm_version
492    return site
```
with
```
488    site = Observatory.get(name)
489    import pint.observatory.special_locations
490    if not isinstance(site,special_locations.BarycenterObs):
491        site.include_gps = include_gps
492        site.include_bipm = include_bipm
493        site.bipm_version = bipm_version
494    return site
```
But this would ignore the arguments passed to get_observatory completely. Perhaps it would even be worth having it through a warning about it.

I tested this change already on my system, and it solves both problems (with TZRMJD and normal barycentered TOAs).

Reply:
I think it will be better to make this more general. We can change the default argument values for include_bipm and include_gps to None, and (re-)set these values only if they are not None.

Reply:
Thanks. If I set defaults to None and reset them only if they are provided, as you suggest, it doesn't actually work, because in pint.toa get_observatory() gets called with explicit arguments:
```
2206         for obs, grp in self.get_obs_groups():
2207             site = get_observatory(
2208                 obs,
2209                 include_gps=include_gps,
2210                 include_bipm=include_bipm,
2211                 bipm_version=bipm_version,
2212             )
``` 
Any idea?

Reply:
I meant changing the `get_observatory` function itself to something like this:

```
def get_observatory(
    name, include_gps=True, include_bipm=True, bipm_version=bipm_default
):
    """Convenience function to get observatory object with options.

    This function will simply call the ``Observatory.get`` method but
    will manually modify the global observatory object after the method is called.
    Name-matching is case-insensitive.

    If the observatory is not present in the PINT list, will fallback to astropy.

    Parameters
    ----------
    name : str
        The name of the observatory
    include_gps : bool, optional
        Set False to disable UTC(GPS)->UTC clock correction.
    include_bipm : bool, optional
        Set False to disable TAI TT(BIPM) clock correction.
    bipm_version : str, optional
        Set the version of TT BIPM clock correction files.

    .. note:: This function can and should be expanded if more clock
        file switches/options are added at a public API level.

    """
    site = Observatory.get(name)
    if include_gps is not None:
        site.include_gps = include_gps
    if include_bipm is not None:
        site.include_bipm = include_bipm
        site.bipm_version = bipm_version
    return site
```

Reply:
Hi,
Yes. That is precisely what I did. It **still fails**, even if I set the defaults of get_observatory to None. This is because toa.py:get_TOAs() sets include_bipm and include_gps to True even when it is called without them (and should therefore default to None), if all TOAs provided do not have yet the clock corrections (as is the case):
```
 283     if all("clkcorr" not in f for f in t.table["flags"]):
 284         if include_gps is None:
 285             include_gps = True
 286         if bipm_version is None:
 287             bipm_version = bipm_default
 288         if include_bipm is None:
 289             include_bipm = True
 290         # FIXME: should we permit existing clkcorr flags?
 291         t.apply_clock_corrections(
 292             include_gps=include_gps,
 293             include_bipm=include_bipm,
 294             bipm_version=bipm_version,
 295             limits=limits,
 296         )
```

The conditions in lines 284 and 288 are met by default, i.e. when calling get_TOAs without explicitly setting include_gps and include_bipm, and those parameters are passed over to apply_clock_corrections, which in turn passes them to get_observatory. Therefore, even if a TOA in a tim file is already defined in the barycenter, whenever it is read with get_TOAs it will get the clock corrections, unless the user manually sets get_TOAs(..., include_gps=False, include_bipm=False). This would, however, affect all TOAs in the tim file, not only the barycentered ones!

The behavior with TZRMJD when TZRSITE is BAT follows from the same origin. While building the TZR TOA, get_TZR_toa(self,toas) in absolute_phase.py reads the clock correction information from the provided toas, and calls get_TOAs_array() explicitly, passing the include_gps and include_bipm parameters that were applied on the toas. Since these were set to True when reading them for the first time, it ends up applying the correction also on TZRMJD, even if it is defined in BAT..

At least, this was my attempt to follow the code around. It involves an interplay of get_observatory() [observatory/__init__.py], get_TOAs(), get_TOAs_array(), apply_clock_corrections.py [toa.py], and get_TZR_toa.py [absolute_phase.py].

Reply:
Hi @abhisrkckl ,

Unfortunately I am on a technical shift to the telescopes in La Palma, I don't have much time these days to continue digging on this. Did you have a chance to have a look at the problem?

So far, the only solution that worked consistently is adding a check for the BarycenterObs class in get_observatory() (https://github.com/nanograv/PINT/pull/1707#discussion_r1440816152), and it passes the tests I've implemented. but it also has some drawbacks. Namely, if apply_clock_corrections() is called on some TOAs multiple times, and there is a barycentered TOA in the list, all calls after the first one will fail.

Cheers,
Giovanni